File: matrix1.c
//without multithreading
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

// Function declarations
int** generateRandomMatrix(int rows, int cols);
int** multiplyMatrices(int** A, int** B, int rowsA, int colsA, int rowsB, int colsB);
void printMatrix(int** matrix, int rows, int cols);

int main() {
    srand(time(NULL)); // Seed for random number generation

    int rowsA, colsA, rowsB, colsB;

    printf("Enter the number of rows for Matrix A: ");
    scanf("%d", &rowsA);
    printf("Enter the number of columns for Matrix A: ");
    scanf("%d", &colsA);

    printf("Enter the number of rows for Matrix B: ");
    scanf("%d", &rowsB);
    printf("Enter the number of columns for Matrix B: ");
    scanf("%d", &colsB);

    if (colsA != rowsB) {
        printf("Matrix multiplication is not possible with the provided dimensions.\n");
        return 1;
    }

    int** matrixA = generateRandomMatrix(rowsA, colsA);
    int** matrixB = generateRandomMatrix(rowsB, colsB);

    int** result = multiplyMatrices(matrixA, matrixB, rowsA, colsA, rowsB, colsB);

    // Print the matrices and the result
    //printf("Matrix A:\n");
    //printMatrix(matrixA, rowsA, colsA);
    //printf("Matrix B:\n");
    //printMatrix(matrixB, rowsB, colsB);
    //printf("Result Matrix:\n");
    //printMatrix(result, rowsA, colsB);

    // Free allocated memory
    free(matrixA[0]);
    free(matrixA);
    free(matrixB[0]);
    free(matrixB);
    free(result[0]);
    free(result);

    return 0;
}

int** generateRandomMatrix(int rows, int cols) {
    int** matrix = (int**)malloc(rows * sizeof(int*));
    for (int i = 0; i < rows; i++) {
        matrix[i] = (int*)malloc(cols * sizeof(int));
        for (int j = 0; j < cols; j++) {
            matrix[i][j] = rand() % 10; // Generates random integers between 0 and 9
        }
    }
    return matrix;
}

int** multiplyMatrices(int** A, int** B, int rowsA, int colsA, int rowsB, int colsB) {
    clock_t startTime = clock();

    int** result = (int**)malloc(rowsA * sizeof(int*));
    for (int i = 0; i < rowsA; i++) {
        result[i] = (int*)malloc(colsB * sizeof(int));
        for (int j = 0; j < colsB; j++) {
            int sum = 0;
            for (int k = 0; k < colsA; k++) {
                sum += A[i][k] * B[k][j];
            }
            result[i][j] = sum;
        }
    }

    clock_t endTime = clock();
    double duration = ((double)(endTime - startTime)) / CLOCKS_PER_SEC * 1000; // Convert to milliseconds
    printf("time taken: %f ms\n", duration);

    return result;
}

void printMatrix(int** matrix, int rows, int cols) {
    for (int i = 0; i < rows; i++) {
        for (int j = 0; j < cols; j++) {
            printf("%d ", matrix[i][j]);
        }
        printf("\n");
    }
}

File: Matrix1.java
//without multithreading

import java.util.Scanner;
import java.util.Random;

public class Matrix1 {
        public static void main(String[] args) {
            Scanner scanner = new Scanner(System.in);

            System.out.print("Enter the number of rows for Matrix A: ");
            int rowsA = scanner.nextInt();
            System.out.print("Enter the number of columns for Matrix A: ");
            int colsA = scanner.nextInt();

            System.out.print("Enter the number of rows for Matrix B: ");
            int rowsB = scanner.nextInt();
            System.out.print("Enter the number of columns for Matrix B: ");
            int colsB = scanner.nextInt();

            if (colsA != rowsB) {
                System.out.println("Matrix multiplication is not possible with the provided dimensions.");
                return;
            }

        int[][] matrixA = generateRandomMatrix(rowsA, colsA);
        int[][] matrixB = generateRandomMatrix(rowsB, colsB);

        int[][] result = multiplyMatrices(matrixA, matrixB);

        // Print the matrices and the result
        System.out.println("Matrix A:");
        //printMatrix(matrixA);
        System.out.println("Matrix B:");
        //printMatrix(matrixB);
        System.out.println("Result Matrix:");
       // printMatrix(result);

    }

    public static int[][] generateRandomMatrix(int rows, int cols) {
        Random random = new Random();
        int[][] matrix = new int[rows][cols];
        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                matrix[i][j] = random.nextInt(10); // Generates random integers between 0 and 9
            }
        }
        return matrix;
    }

    public static int[][] multiplyMatrices(int[][] A, int[][] B) {

        long startTime = System.nanoTime();

        int numRowsA = A.length;
        int numColsA = A[0].length;
        int numRowsB = B.length;
        int numColsB = B[0].length;

        int[][] result = new int[numRowsA][numColsB];

        for (int i = 0; i < numRowsA; i++) {
            for (int j = 0; j < numColsB; j++) {
                int sum = 0;
                for (int k = 0; k < numColsA; k++) {
                    sum += A[i][k] * B[k][j];
                }
                result[i][j] = sum;
            }
        }

        long endTime = System.nanoTime();
        long duration = (endTime - startTime) / 1_000_000; // Convert to milliseconds
        System.out.println("time taken :" + duration);
        return result;
    }

    public static void printMatrix(int[][] matrix) {
        for (int i = 0; i < matrix.length; i++) {
            for (int j = 0; j < matrix[0].length; j++) {
                System.out.print(matrix[i][j] + " ");
            }
            System.out.println();
        }
    }
}
